### PR TITLE
Add Support for Check Runs

### PR DESCRIPTION
**Pull request by dhruvmanila**
Support for Check Runs: https://docs.github.com/en/free-pro-team@latest/rest/reference/checks#check-runs

Added the following endpoints:
- Create a check run
- Update a check run
- Get a check run
- List check run annotations
- List check runs for a Git reference

***Note: 'List check runs in a check suite' endpoint is accessed from `check-suites` which should be applied in #1646 after merging this PR.***

Most of the tests were performed in [this repository](https://www.github.com/dhruvmanila/pygithub-testing) with the help of [this GitHub App](https://github.com/apps/test-pygithub). You can take a look at all the tests performed:

- Add basic check run: https://github.com/dhruvmanila/pygithub-testing/runs/1296028076
- Add completed check run: https://github.com/dhruvmanila/pygithub-testing/runs/1296027873
- Edit check run to success: https://github.com/dhruvmanila/pygithub-testing/runs/1296029552
- Edit check run to failure: https://github.com/dhruvmanila/pygithub-testing/runs/1296029378
- Edit all parameters: https://github.com/dhruvmanila/pygithub-testing/runs/1279259090
- List all annotations for a check run: https://github.com/dhruvmanila/pygithub-testing/runs/1280914700

### CODE REVIEW

**Comment by dhruvmanila** [comment]
Hi, can anyone take a look at this? @s-t-e-v-e-n-k @sfdye 
Thank you if you do :)

**Comment by dstarner** [comment]
Hey, this is a pretty cool feature that we'd like to use, any update on this? (I'm going to try to help bump it @dhruvmanila lol)

@s-t-e-v-e-n-k @sfdye ?

**Comment by sfdye** [comment]
Will review this week

File: github/CheckRun.py [hunk 1]
```diff
@@ -0,0 +1,326 @@
...
+
+    def __repr__(self):
+        return self.get__repr__(
+            {"id": self._id.value, "conclusion": self._conclusion.value}
+        )
+
+    @property
+    def app(self):
+        """
+        :rtype: :class:`github.GithubApp.GithubApp`
```
Comment (s-t-e-v-e-n-k) [inline github/CheckRun.py:49]: Properties have a type, not an rtype.

File: github/CheckRun.py [hunk 2]
```diff
@@ -0,0 +1,326 @@
...
+        """
+        self._completeIfNotSet(self._app)
+        return self._app.value
+
+    @property
+    def check_suite(self):
+        """
+        :rtype: dict
```
Comment (s-t-e-v-e-n-k) [inline github/CheckRun.py:57]: Surely this is a type of CheckSuite? May require another PR to be merged first.

**Review by s-t-e-v-e-n-k — CHANGES_REQUESTED** [review]
This is looking like a great start, I have a few small concerns.

File: github/CheckRun.py (hunk 1 above)
Comment (dhruvmanila) [reply github/CheckRun.py:49]: Nice timing lol. I was just going to reply.
Yes, I will change that.

File: github/CheckRun.py (hunk 2 above)
Comment (dhruvmanila) [reply github/CheckRun.py:57]: I thought the same but then I looked at the payload and we are just getting a dictionary with one attribute in it: `id` of the CheckSuite.
```
...
},
  "name": "mighty_readme",
  "check_suite": {
    "id": 5
  },
  "app": {
  ...
```
Reference: https://docs.github.com/en/free-pro-team@latest/rest/reference/checks#get-a-check-run

Comment (s-t-e-v-e-n-k) [reply github/CheckRun.py]: Then we should probably unwrap it and call the property check_suite_id with type int

Comment (dhruvmanila) [reply github/CheckRun.py]: Done 👍

**Comment by dhruvmanila** [comment]
@sfdye Should I fill the copyrights and license section of `repository.py` and `commit.py` with my info? I tried running `scripts/fix_headers.py` but the script keeps failing.

**Comment by sfdye** [comment]
> @sfdye Should I fill the copyrights and license section of `repository.py` and `commit.py` with my info? I tried running `scripts/fix_headers.py` but the script keeps failing.

Those are not compulsory

**Comment by s-t-e-v-e-n-k** [comment]
I do wonder if we want to move all that stuff into a CONTRIBUTORS file, and drop it from the headers, but that is not for here. In any case, this is looking fantastic.

**Review by s-t-e-v-e-n-k — APPROVED** [review]

**Comment by dhruvmanila** [comment]
Thank you! 

I wonder why the tests failed on push. 🤔

**Comment by s-t-e-v-e-n-k** [comment]
Need to fix the test suite first -- I'm not sure what is going on with httpretty (or even if it's actually the cause, but it's implicated at least)

**Comment by dhruvmanila** [comment]
I think it is httpretty. According to their repository they only support Python version 3.6 and 3.7. I ran tox locally on 3.6 and the tests are passing but it fails on 3.8.

**Comment by s-t-e-v-e-n-k** [comment]
When I run it on Python 3.6 locally, it fails for me ...

**Comment by sbesson** [comment]
Having quickly looked at this, it seems that the issue is related to the latest release of [requests 2.25](https://pypi.org/project/requests/2.25.0/) (which itself pulls `urllib3==1.26`).
Capping `requests` in 

```diff
diff --git a/test-requirements.txt b/test-requirements.txt
index f95cf479..781c3b9b 100644
--- a/test-requirements.txt
+++ b/test-requirements.txt
@@ -1,4 +1,5 @@
 cryptography
+requests<2.25
 httpretty>=0.9.6
 pytest>=5.3
 pytest-cov>=2.8
```

suffices for fixing `tox -e py37` locally for me. 

I haven't looked into the details of the breaking changes in `requests/urllib3` but I assume something in the test infrastructure depends on it.

**Comment by sbesson** [comment]
After further debugging, it looks like this is another occurrence of https://github.com/gabrielfalcao/HTTPretty/issues/409

**Comment by dhruvmanila** [comment]
@sbesson Yes, you are correct. 3.8 and 3.9 also passed by capping requests to < 2.25.

**Comment by s-t-e-v-e-n-k** [comment]
@sbesson and @dhruvmanila Thanks for the assist! I've pushed up https://github.com/PyGithub/PyGithub/pull/1757

**Comment by sbesson** [comment]
@dhruvmanila following up on https://github.com/PyGithub/PyGithub/pull/1727#issuecomment-730109579, are you still looking into updating the check-suite endpoint? Else I might try and find some time this week minimally to fix the conflicts on #1646.

@s-t-e-v-e-n-k: do you have a timeline for releasing a new version of PyGithub with the check API endpoints? From our side, the upcoming end of travis-ci.org is forcing the migration to GitHub workflows and many components of our pipelines are depending on this API. We can definitely use workarounds including temporarily forking this library but being able to consume the upstream library is still our preference if possible at all.

**Comment by sfdye** [comment]
We can make a release within this week, if that is blocking you.

**Comment by dhruvmanila** [comment]
@sfdye If I have permission to work on top of #1646, I will be able to complete it by the weekend. Two endpoints have been completed already in the PR.

**Comment by sbesson** [comment]
> We can make a release within this week, if that is blocking you.

@sfdye in its present state, that would not be sufficient as we would like to consume the API added in #1646. It looks like either @dhruvmanila or myself will be pushing  to get this in a mergeable state. Once integrated, we would certainly be interested in a release of PyGithub.

**Pull request merged**